test/fake_cipherprov.c

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * https://www.openssl.org/source/license.html
+ * or in the file LICENSE in the source distribution.
+ */
+
+#include <string.h>
+#include <openssl/core_names.h>
+#include <openssl/core_object.h>
+#include <openssl/rand.h>
+#include <openssl/provider.h>
+#include <openssl/proverr.h>
+#include <openssl/param_build.h>
+#include "testutil.h"
+#include "fake_cipherprov.h"
+
+#define MAX_KEYNAME 32
+#define FAKE_KEY_LEN 16
+
+typedef struct prov_cipher_fake_ctx_st {
+    char key_name[MAX_KEYNAME];
+    unsigned char key[FAKE_KEY_LEN];
+} PROV_CIPHER_FAKE_CTX;
+
+static int ctx_from_key_params(PROV_CIPHER_FAKE_CTX *pctx, const OSSL_PARAM *params)
+{
+    const OSSL_PARAM *p;
+    char key_name[MAX_KEYNAME];
+    char *pval = key_name;
+
+    memset(key_name, 0, MAX_KEYNAME);
+
+    p = OSSL_PARAM_locate_const(params, FAKE_CIPHER_PARAM_KEY_NAME);
+    if (p != NULL) {
+        if (!OSSL_PARAM_get_utf8_string(p, &pval, sizeof(key_name))) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
+            return 0;
+        }
+
+        memcpy(pctx->key_name, key_name, sizeof(key_name));
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_SKEY_PARAM_RAW_BYTES);
+    if (p != NULL) {
+        size_t data_size = p->data_size;
+
+        if (data_size > FAKE_KEY_LEN)
+            data_size = FAKE_KEY_LEN;
```

Comment:
I know its just a fake provider, but do you also want to store the true length of the key when the param was set, so that you don't return garbage data when exporting the param here?

Reply:
It can be done but doesn't matter from testing perspective

Reply:
the fake provider has a fixed key size, if you try to store a smaller value the rest is random if you try to store a larger value it gets truncated.

Reply:
I think this is resolved, feel free to unresolve if you disagree